### PR TITLE
Improve builtin instantiation

### PR DESCRIPTION
**Pull request by LaszloLango**
* Resolve linker errors with -O0 in some compilers
 * Reduce the stack usage

JerryScript-DCO-1.0-Signed-off-by: László Langó llango.u-szeged@partner.samsung.com

### CODE REVIEW

**Comment by LaszloLango** [comment]
Benchmark | Stack (bytes) | Perf (sec) |
----: | ----: | ----: | 
3d-cube.js | 18960 -> 15016 : +20.802% | 0.830 -> 0.835 : -0.623% | 
3d-raytrace.js | 3220 -> 2716 : +15.652% | 1.083 -> 1.082 : +0.115% | 
access-binary-trees.js | 3928 -> 3208 : +18.330% | 0.586 -> 0.588 : -0.302% | 
access-fannkuch.js | 1472 -> 1472 : 0.000% | 2.149 -> 2.158 : -0.455% | 
access-nbody.js | 1592 -> 1432 : +10.050% | 1.189 -> 1.195 : -0.481% | 
bitops-3bit-bits-in-byte.js | 1432 -> 1432 : 0.000% | 0.538 -> 0.535 : +0.495% | 
bitops-bits-in-byte.js | 1432 -> 1432 : 0.000% | 0.737 -> 0.737 : -0.035% | 
bitops-bitwise-and.js | 1208 -> 1208 : 0.000% | 0.951 -> 0.945 : +0.611% | 
bitops-nsieve-bits.js | 1432 -> 1432 : 0.000% | 1.231 -> 1.241 : -0.804% | 
controlflow-recursive.js | 86112 -> 67752 : +21.321% | 0.397 -> 0.399 : -0.534% | 
crypto-aes.js | 2164 -> 2124 : +1.848% | 0.914 -> 0.926 : -1.349% | 
crypto-md5.js | 2204 -> 1844 : +16.334% | 0.649 -> 0.652 : -0.407% | 
crypto-sha1.js | 1516 -> 1504 : +0.792% | 0.647 -> 0.646 : +0.086% | 
date-format-tofte.js | 2120 -> 1904 : +10.189% | 0.730 -> 0.731 : -0.174% | 
date-format-xparb.js | 2536 -> 2320 : +8.517% | 0.431 -> 0.409 : +5.054% | 
math-cordic.js | 1508 -> 1432 : +5.040% | 1.271 -> 1.275 : -0.245% | 
math-partial-sums.js | 1376 -> 1376 : 0.000% | 0.754 -> 0.763 : -1.250% | 
math-spectral-norm.js | 1828 -> 1540 : +15.755% | 0.582 -> 0.575 : +1.067% | 
string-base64.js | 1432 -> 1432 : 0.000% | 1.583 -> 1.553 : +1.904% | 
string-fasta.js | 1416 -> 1416 : 0.000% | 1.368 -> 1.363 : +0.331% | 
Geometric mean: | +7.578% | +0.159% | 

Binary sizes (bytes)
e10f6d6adf:135004
e8aadcdfe2:134684

File: jerry-core/ecma/builtin-objects/ecma-builtins.c [hunk 1]
```diff
@@ -39,18 +39,90 @@ static void ecma_instantiate_builtin (ecma_builtin_id_t id);
  */
 typedef const ecma_builtin_property_descriptor_t *ecma_builtin_property_list_reference_t;
 
+typedef ecma_value_t (*ecma_builtin_dispatch_routine_t)(uint16_t builtin_routine_id,
+                                              ecma_value_t this_arg,
```
Comment (galpeter) [inline jerry-core/ecma/builtin-objects/ecma-builtins.c]: The indentations are incorrect here.

File: jerry-core/ecma/builtin-objects/ecma-builtins.c [hunk 2]
```diff
@@ -39,18 +39,90 @@ static void ecma_instantiate_builtin (ecma_builtin_id_t id);
...
+  #include "ecma-builtins.inc.h"
+  #undef BUILTIN
+  #undef BUILTIN_CALLABLE
+};
+
+static const ecma_builtin_dispatch_call_t ecma_builtin_call_functions[] =
+{
+  #define BUILTIN(a, b, c, d, e)
+  #define BUILTIN_CALLABLE(builtin_id, \
+                  object_type, \
```
Comment (galpeter) [inline jerry-core/ecma/builtin-objects/ecma-builtins.c]: Indentation is incorrect here.

File: jerry-core/ecma/builtin-objects/ecma-builtins.c [hunk 3]
```diff
@@ -39,18 +39,90 @@ static void ecma_instantiate_builtin (ecma_builtin_id_t id);
...
+  #include "ecma-builtins.inc.h"
+  #undef BUILTIN_CALLABLE
+  #undef BUILTIN
+};
+
+static const ecma_builtin_dispatch_call_t ecma_builtin_construct_functions[] =
+{
+  #define BUILTIN(a, b, c, d, e)
+  #define BUILTIN_CALLABLE(builtin_id, \
+                  object_type, \
```
Comment (galpeter) [inline jerry-core/ecma/builtin-objects/ecma-builtins.c]: ditto

File: jerry-core/ecma/builtin-objects/ecma-builtins.c [hunk 4]
```diff
@@ -39,18 +39,90 @@ static void ecma_instantiate_builtin (ecma_builtin_id_t id);
...
+};
+
 /**
  * Property descriptor lists for all built-ins.
  */
 static const ecma_builtin_property_list_reference_t ecma_builtin_property_list_references[] =
 {
+#define BUILTIN(a, b, c, d, e)
+#define BUILTIN_CALLABLE(builtin_id, \
+                object_type, \
```
Comment (galpeter) [inline jerry-core/ecma/builtin-objects/ecma-builtins.c]: ditto

File: jerry-core/ecma/builtin-objects/ecma-builtins.c [hunk 5]
```diff
@@ -331,6 +402,24 @@ ecma_instantiate_builtin (ecma_builtin_id_t id) /**< built-in id */
 {
   switch (id)
   {
+#define BUILTIN(a, b, c, d, e)
+#define BUILTIN_CALLABLE(builtin_id, \
```
Comment (galpeter) [inline jerry-core/ecma/builtin-objects/ecma-builtins.c]: Can't we just call the other define in the BUILTIN_CALLABLE here?
ex.:
```
#define BUILTIN_CALLABLE(a) BUILTIN(a)
```

File: jerry-core/ecma/builtin-objects/ecma-builtins.h [hunk 6]
```diff
@@ -23,13 +23,26 @@
  */
 typedef enum
 {
+#define BUILTIN(a, b, c, d, e)
+#define BUILTIN_CALLABLE(builtin_id, \
+                object_type, \
```
Comment (galpeter) [inline jerry-core/ecma/builtin-objects/ecma-builtins.h]: Incorrect indention.

File: jerry-core/ecma/builtin-objects/ecma-builtins.h [hunk 7]
```diff
@@ -23,13 +23,26 @@
...
+                object_prototype_builtin_id, \
+                is_extensible, \
+                lowercase_name) \
+  builtin_id,
```
Comment (galpeter) [inline jerry-core/ecma/builtin-objects/ecma-builtins.h]: Is this separation required somewhere? That is do we expect the order of the builtin enums somewhere?

File: jerry-core/ecma/builtin-objects/ecma-builtins.inc.h [hunk 8]
```diff
@@ -24,7 +24,7 @@ BUILTIN (ECMA_BUILTIN_ID_OBJECT_PROTOTYPE,
          object_prototype)
 
 /* The Object object (15.2.1) */
-BUILTIN (ECMA_BUILTIN_ID_OBJECT,
+BUILTIN_CALLABLE (ECMA_BUILTIN_ID_OBJECT,
          ECMA_OBJECT_TYPE_FUNCTION,
```
Comment (galpeter) [inline jerry-core/ecma/builtin-objects/ecma-builtins.inc.h]: The indentations are incorrect in this file.

File: jerry-core/ecma/builtin-objects/ecma-builtins-internal.h [hunk 9]
```diff
@@ -69,7 +69,7 @@ typedef struct
   uint16_t value; /**< value of the property */
 } ecma_builtin_property_descriptor_t;
 
-#define BUILTIN(builtin_id, \
+#define BUILTIN_CALLABLE(builtin_id, \
```
Comment (zherczeg) [inline jerry-core/ecma/builtin-objects/ecma-builtins-internal.h]: What about BUILTIN_CALLABLE -> BUILTIN_ROUTINE ?

File: jerry-core/ecma/builtin-objects/ecma-builtins.h (hunk 7 above)
Comment (LaszloLango) [reply jerry-core/ecma/builtin-objects/ecma-builtins.h]: Yes, it is necessary to be in that order.

File: jerry-core/ecma/builtin-objects/ecma-builtins.c (hunk 5 above)
Comment (LaszloLango) [reply jerry-core/ecma/builtin-objects/ecma-builtins.c]: Good idea

**Pull request merged**